### PR TITLE
cmd/syncthing, lib/relay: Fixes regarding stopping of services

### PR DESCRIPTION
**Pull request by imsodin**
While running  #5216 with race detector on my laptop ("in production") I got three races:

```
==================
WARNING: DATA RACE
Read at 0x00c002377ab0 by goroutine 338:
  github.com/syncthing/syncthing/lib/relay/client.(*staticClient).Stop()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/relay/client/static.go:172 +0x47
  github.com/syncthing/syncthing/lib/relay/client.(*dynamicClient).Stop()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/relay/client/dynamic.go:128 +0x109
  github.com/syncthing/syncthing/lib/connections.(*relayListener).Stop()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/connections/relay_listen.go:124 +0xac
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).stopSupervisor.func1()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:590 +0x3e
Previous write at 0x00c002377ab0 by goroutine 65:
  [failed to restore the stack]
Goroutine 338 (running) created at:
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).stopSupervisor()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:589 +0x2d0
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).Serve()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:434 +0x108b
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).runService.func1()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:540 +0x6c
Goroutine 65 (running) created at:
  github.com/syncthing/syncthing/lib/connections.(*relayListener).Serve()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/connections/relay_listen.go:57 +0x1f3
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).runService.func1()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:540 +0x6c
==================
WARNING: DATA RACE
Read at 0x00c002377ab8 by goroutine 338:
  github.com/syncthing/syncthing/lib/relay/client.(*staticClient).Stop()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/relay/client/static.go:177 +0x85
  github.com/syncthing/syncthing/lib/relay/client.(*dynamicClient).Stop()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/relay/client/dynamic.go:128 +0x109
  github.com/syncthing/syncthing/lib/connections.(*relayListener).Stop()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/connections/relay_listen.go:124 +0xac
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).stopSupervisor.func1()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:590 +0x3e
Previous write at 0x00c002377ab8 by goroutine 65:
  [failed to restore the stack]
Goroutine 338 (running) created at:
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).stopSupervisor()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:589 +0x2d0
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).Serve()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:434 +0x108b
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).runService.func1()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:540 +0x6c
Goroutine 65 (running) created at:
  github.com/syncthing/syncthing/lib/connections.(*relayListener).Serve()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/connections/relay_listen.go:57 +0x1f3
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).runService.func1()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:540 +0x6c
==================
WARNING: DATA RACE
Read at 0x00c00054c5c0 by goroutine 340:
  main.(*usageReportingService).Stop()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/cmd/syncthing/usage_report.go:396 +0x3e
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).stopSupervisor.func1()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:590 +0x3e
Previous write at 0x00c00054c5c0 by goroutine 108:
  [failed to restore the stack]
Goroutine 340 (running) created at:
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).stopSupervisor()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:589 +0x2d0
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).Serve()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:434 +0x108b
Goroutine 108 (running) created at:
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).runService()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:528 +0x68
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).Serve()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:426 +0xe99
==================
```
Both types have in common, that they have close channels created in both their constructor and `Serve` method. I removed the redundant creations in `Serve` and added `sync.Once` members to make sure multiple calls to `Stop` will close the channel just once (unrelated to the race, but still seems like the right thing to do).

### CODE REVIEW

File: lib/relay/client/static.go [hunk 1]
```diff
@@ -169,11 +169,7 @@ func (c *staticClient) Serve() {
 }
 
 func (c *staticClient) Stop() {
-	if c.stop == nil {
-		return
-	}
-
-	close(c.stop)
+	c.stopOnce.Do(func() { close(c.stop) })
```
Comment (AudriusButkevicius) [inline lib/relay/client/static.go]: Ditto?

Comment (imsodin) [reply lib/relay/client/static.go]: Don't get it.

**Comment by imsodin** [comment]
Actu

**Comment by AudriusButkevicius** [comment]
Lol, git on mobile decided to discard one of my comments. I was saying that your change makes the services non-restartable.

**Pull request closed without merging**